chunky/src/java/se/llbit/chunky/resources/SignTexture.java

```diff
@@ -36,74 +38,84 @@
+          ystart += lineHeight;
+          continue;
         }
-      }
-      int xstart = (int) Math.ceil((width - lineWidth) / 2.0);
-      for (JsonValue textItem : line) {
-        String textLine = textItem.object().get("text").stringValue("");
-        Color color = Color.get(textItem.object().get("color").intValue(0));
+        int lineWidth = 0;
+        for(JsonValue textItem : line) {
+          String textLine = textItem.object().get("text").stringValue("");
+          for(int c : textLine.codePoints().toArray()) {
+            Glyph glyph = Texture.fonts.getGlyph(c);
+            lineWidth += glyph != null ? glyph.width : 0;
+          }
+        }
+        int xstart = (int) Math.ceil((width - lineWidth) / 2.0);
+        for(JsonValue textItem : line) {
+          String textLine = textItem.object().get("text").stringValue("");
+          Color color = Color.get(textItem.object().get("color").intValue(0));
 
-        for (int c : textLine.codePoints().toArray()) {
-          Glyph glyph = Texture.fonts.getGlyph(c);
-          if (glyph != null) {
-            int y = ystart - glyph.ascent + lineHeight;
+          for(int c : textLine.codePoints().toArray()) {
+            Glyph glyph = Texture.fonts.getGlyph(c);
+            if(glyph != null) {
+              int y = ystart - glyph.ascent + lineHeight;
 
-            for (int py = 0; py < glyph.height; ++py) {
-              int x = xstart;
-              for (int px = glyph.xmin; px <= glyph.xmax; ++px) {
-                if ((glyph.lines[py] & (1 << px)) != 0) {
-                  data[y * width + x] = color.rgbColor;
+              for(int py = 0; py < glyph.height; ++py) {
+                int x = xstart;
+                for(int px = glyph.xmin; px <= glyph.xmax; ++px) {
+                  if((glyph.lines[py] & (1 << px)) != 0) {
+                    textColor.setPixel(x, y, color.id);
+                    textMask.setPixel(x, y, true);
+                  }
+                  x += 1;
                 }
-                x += 1;
+                y += 1;
               }
-              y += 1;
+              xstart += glyph.width;
             }
-            xstart += glyph.width;
           }
         }
+        ystart += lineHeight;
       }
-      ystart += lineHeight;
-    }
-    texture = new Texture(img);
-  }
-
-  @Override
-  public void getColor(double u, double v, Vector4 c) {
-    texture.getColor(u, v, c);
-    if (c.w == 0) {
-      signTexture.getColor(u * ww + u0, v * hh + v0, c);
     }
   }
 
   @Override
   public float[] getColor(double u, double v) {
-    float[] rgba = texture.getColor(u, v);
-    if (rgba[3] == 0) {
+    int x = (int)(u * 96 - Ray.EPSILON);
+    int y = (int) ((1 - v) * 48 - Ray.EPSILON);
+    if(textMask != null && textMask.getPixel(x, y)) {
+      float[] result = new float[4];
+      Color characterColor = Color.get(textColor.getPixel(x, y));
+      ColorUtil.getRGBAComponentsGammaCorrected(characterColor.rgbColor, result);
```

Comment:
We could store the gamma-corrected color in `textColor`. Slightly slower while loading but slightly faster while rendering.

Reply:
I guess you mean storing them in `Color` rather? I'll do that (It would even save me some allocation, yeah!)

Reply:
I was thinking about pre-calculating the gamma-corrected values for the texture